src/twisted/python/systemd.py

```diff
@@ -68,26 +70,70 @@ def fromEnvironment(
         if start is None:
             start = cls._START
 
-        descriptors: List[int] = []
-
-        try:
-            pid = int(environ["LISTEN_PID"])
-        except (KeyError, ValueError):
-            pass
+        if str(getpid()) == environ.get("LISTEN_PID"):
+            descriptors: List[int] = _parseDescriptors(start, environ)
+            names: Sequence[str] = _parseNames(environ)
         else:
-            if pid == getpid():
-                try:
-                    count = int(environ["LISTEN_FDS"])
-                except (KeyError, ValueError):
-                    pass
-                else:
-                    descriptors = list(range(start, start + count))
-                    del environ["LISTEN_PID"], environ["LISTEN_FDS"]
-
-        return cls(descriptors)
-
-    def inheritedDescriptors(self) -> Iterable[int]:
+            descriptors = []
+            names = ()
+
+        # They may both be missing (consistent with not running under systemd
+        # at all) or they may both be present (consistent with running under
+        # systemd 227 or newer).  It is not allowed for only one to be present
+        # or for the values to disagree with each other.
+        if len(names) != len(descriptors):
+            return cls([], ())
+
+        return cls(descriptors, names)
+
+    def inheritedDescriptors(self) -> List[int]:
         """
         @return: The configured descriptors.
         """
         return list(self._descriptors)
+
+    def inheritedNamedDescriptors(self) -> Dict[str, int]:
```

Comment:
this is ok as I guess is the same "design" as the previous `inheritedDescriptors` API.

But do we need to get the full dict?

I see this is rather used as `inheritedNamedDescriptors(name) -> int`

But this being a low-level, generic API, maybe the current API design is OK.

Reply:
Yea I think that probably would have been alright but I guess consistency in the interface is better than just doing something marginally better for one of the two APIs.  Indeed it is a low-level component that most applications will never touch.

Reply:
(I guess it's also worth mentioning that with the list/dict interfaces you get strictly more power since they let you enumerate the descriptors and you can always index them yourself ... I don't know what usage pattern would actually benefit from this but maybe there is one)